modules/post/linux/gather/enum_containers.rb

```diff
@@ -0,0 +1,94 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Post
+  include Msf::Post::File
+
+  def initialize(info = {})
+    super(update_info(info,
+      'Name' => 'Linux Container Enumeration',
+      'Description' => %q{
+        This module attempts to enumerate containers running on the target machine.
+        Supports Docker, LXC and RKT.
+      },
+      'License' => MSF_LICENSE,
+      'Author' => ['Mat Rollings [stealthcopter]'],
+      'Platform' => ['linux'],
+      'SessionTypes' => ['shell', 'meterpreter']
+    ))
+  end
+
+  # Check if a container program is installed and usable by current user
+  def runnable(container_type)
+    case container_type
+    when 'docker'
+      result = cmd_exec('docker ps >/dev/null 2>&1 && echo true')
+    when 'lxc'
+      result = cmd_exec('lxc list >/dev/null 2>&1 && echo true')
+    when 'rkt'
+      result = cmd_exec('rkt list >/dev/null 2>&1 && echo true')
+    else
+      print_error("Invalid container type #{container_type}")
+      return false
+    end
+    result =~ /true$/
+  end
+
+  # Count the number of currently running containers
+  def count_containers(container_type)
+    case container_type
+    when 'docker'
+      result = cmd_exec('docker ps --format "{{.Names}}" 2>/dev/null | wc -l')
+    when 'lxc'
+      result = cmd_exec('lxc list -c n --format csv 2>/dev/null | wc -l')
+    when 'rkt'
+      result = cmd_exec('rkt list 2>/dev/null | tail -n +2  | wc -l')
+    else
+      print_error("Invalid container type '#{container_type}'")
+      return 0
+    end
+    result.to_i
+  end
+
+  # List the currently running containers
+  def list_containers(container_type)
+    case container_type
+    when 'docker'
+      result = cmd_exec('docker ps')
```

Comment:
As an aside from the above question of `docker ps` potentially being deprecated:

What are your thoughts on changing this to `docker ps -a` - that way the user can see can see all previously ran containers - which could be theoretically be run again to extract creds etc.

Using a concrete example, for me locally there are no results for `docker ps`

```
$ docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
```

But if I run with the `-a` option:

```
Options:
  -a, --all             Show all containers (default shows just running)
```

This gives me access to containers that I've ran in the past. These can all be spun up again:

```
$ docker ps -a
CONTAINER ID        IMAGE                                             COMMAND                  CREATED             STATUS                            PORTS                    NAMES
08538f6366c8        metasploitframework/metasploit-framework:latest   "docker/entrypoint.s…"   3 weeks ago         Exited (137) About a minute ago                            metasploit-framework_metasploit_1
95d158abfef5        metasploit:dev                                    "docker/entrypoint.s…"   3 weeks ago         Exited (1) 3 weeks ago                                     metasploit-framework_ms_1
83e6b7908bf4        postgres:10-alpine                                "docker-entrypoint.s…"   3 weeks ago         Exited (255) 3 weeks ago          0.0.0.0:5432->5432/tcp   metasploit-framework_db_1
1bca93638977        kalilinux/kali-rolling                            "/bin/bash"              3 weeks ago         Exited (127) 3 weeks ago                                   competent_hermann
63512d5c8be8        kalilinux/kali-rolling                            "/bin/bash"              4 weeks ago         Exited (130) 4 weeks ago                                   reverent_curie

```

Now we can grab the container ID, start it again, and create a shell to start running commands on:

```
$ docker start 08538f6366c8
08538f6366c8
$ docker exec -it 08538f6366c8 /bin/bash
bash-5.0# whoami
root
➜  metasploit-framework docker exec -it 08538f6366c8 /bin/bash
bash-5.0# whoami
root
bash-5.0# env
DATABASE_URL=postgres://postgres:msfpassword@db:5432/msf <--- Oops.
HOSTNAME=08538f6366c8
RUBY_DOWNLOAD_SHA256=d5d6da717fd48524596f9b78ac5a2eeb9691753da5c06923a6c31190abe01a62
NMAP_PRIVILEGED=
RUBY_VERSION=2.6.5
PWD=/usr/src/metasploit-framework
BUNDLE_APP_CONFIG=/usr/local/bundle
RUBY_MAJOR=2.6
APP_HOME=/usr/src/metasploit-framework
HOME=/root
METASPLOIT_GROUP=metasploit
BUNDLE_SILENCE_ROOT_WARNING=1
GEM_HOME=/usr/local/bundle
TERM=xterm
SHLVL=1
PATH=/usr/local/bundle/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
_=/usr/bin/env
```

In this hypothetical scenario i've just successfully extracted a `DATABASE_URL` which contains direct access to a database which could contain sensitive data, further pivoting capabilities, or a reused password 😱

Reply:
Yeah I agree. I think the best option is probably to add a Boolean option to the module that sets whether to list inactive containers. I'll just need to make sure the other container commands are also aligned.

Reply:
Resolved: now all platforms show all containers and have a count of running and total containers. Also added a `CMD` arg based on above that can run a command on all containers found and run a command such as `env` in order to enumerate further.